sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1237,20 +1250,48 @@ class Analyzer(
       if (conflictPlans.isEmpty) {
         right
       } else {
-        val attributeRewrites = AttributeMap(conflictPlans.flatMap {
-          case (oldRelation, newRelation) => oldRelation.output.zip(newRelation.output)})
-        val conflictPlanMap = conflictPlans.toMap
-        // transformDown so that we can replace all the old Relations in one turn due to
-        // the reason that `conflictPlans` are also collected in pre-order.
-        right transformDown {
-          case r => conflictPlanMap.getOrElse(r, r)
-        } transformUp {
-          case other => other transformExpressions {
+        rewritePlan(right, conflictPlans.toMap)._1
+      }
+    }
+
+    private def rewritePlan(plan: LogicalPlan, conflictPlanMap: Map[LogicalPlan, LogicalPlan])
+      : (LogicalPlan, Seq[(Attribute, Attribute)]) = {
+      if (conflictPlanMap.contains(plan)) {
+        // If the plan is the one that conflict the with left one, we'd
+        // just replace it with the new plan and collect the rewrite
+        // attributes for the parent node.
+        val newRelation = conflictPlanMap(plan)
+        newRelation -> plan.output.zip(newRelation.output)
+      } else {
+        val attrMapping = new mutable.ArrayBuffer[(Attribute, Attribute)]()
+        val newPlan = plan.mapChildren { child =>
+          // If not, we'd rewrite child plan recursively until we find the
+          // conflict node or reach the leaf node.
+          val (newChild, childAttrMapping) = rewritePlan(child, conflictPlanMap)
+          attrMapping ++= childAttrMapping.filter { case (oldAttr, _) =>
+            // If an attribute from the child is not output or referenced by the parent plan,
+            // it's useless to propagate this attribute to the parent plan, as they are not
+            // going to use this attribute.
+            (plan.outputSet ++ plan.references).contains(oldAttr)
```

Comment:
the parents of `plan` can only reference `plan.outputSet`, that's why I think only `plan.outputSet` is necessary.

Reply:
ah sorry, you are right. `attrMapping` is not only propagated to parents, but is also used to rewrite attributes of the currennt `plan`, so `plan.references` is necessary. Can we make the comment clearer about it?

Reply:
maybe something like
```
// `attrMapping` is used to replace the attributes of the current `plan`, and also to be propagated to
// parent plans, so the `oldAttr` must either be part of `plan.references` (so that it can be used to
// replace attributes of current plan) or `plan.outputSet` (so that it can be used by parent plans).
```